torch/distributions/bernoulli.py

```diff
@@ -46,6 +46,19 @@ def __init__(self, probs=None, logits=None, validate_args=None):
             batch_shape = self._param.size()
         super(Bernoulli, self).__init__(batch_shape, validate_args=validate_args)
 
+    def expand(self, batch_shape):
+        batch_shape = torch.Size(batch_shape)
+        new = self.__new__(Bernoulli)
```

Comment:
That is unfortunately true. :( I think we can use `type(self)` to ensure that we are getting the correct instance type, but if we have made any customizations to `__init__`, this will require us to also override `.expand` and plug in those changes. That's not ideal. Another thing we can do is check if `type(self)` doesn't match the class and if so call the `__init__` method for safety.

Reply:
Hmm, I think it's too draconian to `assert type(self) is Bernoulli`, since we should allow subclassing w/o overriding the `__init__` method. I think the safest thing to do would be:
```py
if self.__init__.im_func is not Bernoulli.__init__.im_func:
    raise NotImplementedError("Subclasses that define a custom __init__ method "
                              "must also define a custom .expand() method")
```

Reply:
Yes, that would be better. Rather than trying to do something by default, it's better to throw this error since its not currently possible to `.expand` without making any assumptions about `.__init__` (e.g. we may introduce some new parameters that would additionally require broadcasting).

Reply:
The `type(self).__init__ is not Bernoulli.__init__` version seems palatable.

Reply:
Hmm..I think this is getting a bit complicated. Consider a distribution `X` that subclasses `Y`. We may call `super(X, self).__init__(...)` to set attributes defined in the parent class. If we want to use `super(X, self).expand(...)` to expand those attributes, this isn't currently possible (apart from this additional check) since `.expand` not just sets the attributes (like `__init__`) but also creates a new instance via `__new__`.  One way (there may be a better one) might be to take in an optional `instance` arg to enable subclasses to use the `expand` method :

```
    def expand(self, batch_shape, instance=None):
        if not instance and type(self).__init__ is not Bernoulli.__init__:
            raise NotImplementedError("Subclasses that define a custom __init__ method "
                                      "must also define a custom .expand() method")
        new = self.__new__(Bernoulli) if not instance else instance
        ...
```

Reply:
I don't think we need to support generic `.expand()` via `super(X, self).expand()`. In cases where a user wants to subclass with an `__init__` and use `.expand()`, I guess we could require the user to hand-code an `.expand()` method? That's what we're doing with `TransformedDistribution`, right?

Reply:
We are still able to call `base_dist.expand()` in `TransformedDistribution`, so it isn't as bad. For subclasses, we will need to duplicate the `.expand` code from the parent class just to set those attributes. e.g. in `Chi2`, we will need to set `concentration` and `rate` in `.expand`, which are not used inside `Chi2`. I'm not sure whether we should provide this extra `instance` kwarg for convenience, or leave it to the users to hand-code it.